concurrency/DocumentCommitThread.java

```diff
@@ -0,0 +1,953 @@
+  }
+
+  private void cancel(@NonNls @Nonnull Object reason) {
+    startNewTask(null, reason);
+  }
+
+  @Override
+  public void commitAsynchronously(@Nonnull final Project project,
+                                   @Nonnull final Document document,
+                                   @NonNls @Nonnull Object reason,
+                                   @Nullable TransactionId context) {
+    assert !isDisposed : "already disposed";
+
+    if (!project.isInitialized()) return;
+    PsiFile psiFile = PsiDocumentManager.getInstance(project).getCachedPsiFile(document);
+    if (psiFile == null || psiFile instanceof PsiCompiledElement) return;
+    doQueue(project, document, getAllFileNodes(psiFile), reason, context,
+            PsiDocumentManager.getInstance(project).getLastCommittedText(document));
+  }
+
+  private void doQueue(@Nonnull Project project,
+                       @Nonnull Document document,
+                       @Nonnull List<Pair<PsiFileImpl, FileASTNode>> oldFileNodes,
+                       @Nonnull Object reason,
+                       @Nullable TransactionId context,
+                       @Nonnull CharSequence lastCommittedText) {
+    synchronized (lock) {
+      if (!project.isInitialized()) return;  // check the project is disposed under lock.
+      CommitTask newTask = createNewTaskAndCancelSimilar(project, document, oldFileNodes, reason, context,
+                                                         lastCommittedText);
+
+      documentsToCommit.offer(newTask);
+      log(project, "Queued", newTask, reason);
+
+      wakeUpQueue();
+    }
+  }
+
+  @Nonnull
+  private CommitTask createNewTaskAndCancelSimilar(@Nonnull Project project,
+                                                   @Nonnull Document document,
+                                                   @Nonnull List<Pair<PsiFileImpl, FileASTNode>> oldFileNodes,
+                                                   @Nonnull Object reason,
+                                                   @Nullable TransactionId context,
+                                                   @Nonnull CharSequence lastCommittedText) {
+    synchronized (lock) {
+      for (Pair<PsiFileImpl, FileASTNode> pair : oldFileNodes) {
+        assert pair.first.getProject() == project;
+      }
+      CommitTask newTask = new CommitTask(project, document, oldFileNodes, createProgressIndicator(), reason, context,
+                                          lastCommittedText);
+      cancelAndRemoveFromDocsToCommit(newTask, reason);
+      cancelAndRemoveCurrentTask(newTask, reason);
+      cancelAndRemoveFromDocsToApplyInEDT(newTask, reason);
+
+      return newTask;
+    }
+  }
+
+  @SuppressWarnings({"NonConstantStringShouldBeStringBuffer", "StringConcatenationInLoop"})
+  public void log(Project project, @NonNls String msg, @Nullable CommitTask task, @NonNls Object... args) {
+    if (true) return;
+
+    String indent = new SimpleDateFormat("hh:mm:ss:SSSS").format(new Date()) +
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The SimpleDateFormat object was created without setting the time zone. Make sure that you want to use the default time zone or use letter z, Z or X in the pattern. Otherwise, if you do not call setTimeZone() on the created SimpleDateFormat object, you might get an unexpected date and time when using the object.